### ISSUE TITLE
Mesh and shape smoothing - GSoC17

### ISSUE THREAD

**Issue by kkatrio** [CGAL/cgal#2377]
## Overview
This is a brief report on the project *mesh and shape smoothing* which was done in terms of Google summer of code 2017.

The projected aimed at enhancing Polygon Mesh Processing package with two functionalities that offer mesh and shape smoothing. Mesh smoothing improves the quality of the triangle elements using geometric characteristics such as angles and areas, whereas shape smoothing affects the overall shape by smoothing out sharp areas with large curvature without depending directly on the geometric characteristics.

## Code
The basic part of the work is done and is ready to be submitted as a small feature to be evaluated and merged to the master CGAL branch. The work was done [in this cgal branch](https://github.com/CGAL/cgal-public-dev/tree/gsoc2017-smoothing-kkatrio). 

This is a [link to all the commits](https://github.com/CGAL/cgal-public-dev/commits/gsoc2017-smoothing-kkatrio), and here is a list with all the commits:

https://github.com/CGAL/cgal-public-dev/commit/43a057b944 License on evaluation class
https://github.com/CGAL/cgal-public-dev/commit/c42b835f8b mesh quality evaluation class
https://github.com/CGAL/cgal-public-dev/commit/c2f493e0e4 results & images in docs
https://github.com/CGAL/cgal-public-dev/commit/51712c80ea doc image refinement
https://github.com/CGAL/cgal-public-dev/commit/408cfe558a License
https://github.com/CGAL/cgal-public-dev/commit/fd08fba3b7 demo plugin: add iterations spinbox for curvature flow
https://github.com/CGAL/cgal-public-dev/commit/101ee9f118 document overloads
https://github.com/CGAL/cgal-public-dev/commit/c6ea17007c renaming functions in API
https://github.com/CGAL/cgal-public-dev/commit/cf22e2dbfc fix and clean test files
https://github.com/CGAL/cgal-public-dev/commit/fbe1aa58ba images in user manual
https://github.com/CGAL/cgal-public-dev/commit/5a23148585 clean curvature flow impl
https://github.com/CGAL/cgal-public-dev/commit/aa6eb80089 plugin improvement
https://github.com/CGAL/cgal-public-dev/commit/c283d4455f corrections in verbose output
https://github.com/CGAL/cgal-public-dev/commit/9bdf52e059 improved doc
https://github.com/CGAL/cgal-public-dev/commit/70acc67e1b compatible remeshing function
https://github.com/CGAL/cgal-public-dev/commit/c03f5715a0 cleaning code & use weight calculator which covers better at 0 and Pi making possible many curvature iterations
https://github.com/CGAL/cgal-public-dev/commit/5a414561f2 testing different weight calculation methods for curvature flow
https://github.com/CGAL/cgal-public-dev/commit/6579c3c2e5 remove trivial example
https://github.com/CGAL/cgal-public-dev/commit/84d8ae86c0 init after removing degenerate faces
https://github.com/CGAL/cgal-public-dev/commit/f55bd058c0 counting not moved points for debug
https://github.com/CGAL/cgal-public-dev/commit/6f25484d1b angle remeshing only if it impoves min angle
https://github.com/CGAL/cgal-public-dev/commit/f30d9c9260 calculate movement in separate function
https://github.com/CGAL/cgal-public-dev/commit/1be2fc0de6 selected vertices
https://github.com/CGAL/cgal-public-dev/commit/4848460b64 examples
https://github.com/CGAL/cgal-public-dev/commit/9c75d91cd1 cmakelists for smoothing tests
https://github.com/CGAL/cgal-public-dev/commit/1c1e6ebd3c unit tests for curvature flow
https://github.com/CGAL/cgal-public-dev/commit/6f8adc284a unit tests for smoothing
https://github.com/CGAL/cgal-public-dev/commit/8a5f85fe8f user manual doc
https://github.com/CGAL/cgal-public-dev/commit/8cc5d3775b reference manual documentation
https://github.com/CGAL/cgal-public-dev/commit/9ca162840e debugging curvature flow
https://github.com/CGAL/cgal-public-dev/commit/ae2b401edc change directory
https://github.com/CGAL/cgal-public-dev/commit/2725105c83 use constrained vertices on all types of smoothing
https://github.com/CGAL/cgal-public-dev/commit/1ee4dc6633 half-sphere data
https://github.com/CGAL/cgal-public-dev/commit/cc4fbcd485 clean redundant code & minor syntax corrections
https://github.com/CGAL/cgal-public-dev/commit/216de5b70f another attempt to curvature flow
https://github.com/CGAL/cgal-public-dev/commit/07628a546e plugin nb iterations parameter
https://github.com/CGAL/cgal-public-dev/commit/ab4b5997bb improve demo plugin by separating functionalities
https://github.com/CGAL/cgal-public-dev/commit/a442cd8211  use opposite direction for move vector & remove monge jet fitting code
https://github.com/CGAL/cgal-public-dev/commit/d5e4994fef secure cot calculation & avoid division with zero weight
https://github.com/CGAL/cgal-public-dev/commit/2bb90b3197 fixes on weights & degnerate cases
https://github.com/CGAL/cgal-public-dev/commit/abcd400f41 fix compilation
https://github.com/CGAL/cgal-public-dev/commit/5a7506ad68 fix compilation of smoothing_plugin
https://github.com/CGAL/cgal-public-dev/commit/017030ba8e fix bug with curvature flow - sphere test is much better
https://github.com/CGAL/cgal-public-dev/commit/102fcfe94a all degenerate faces are taken out with remove_degenerate_faces - for now at least
https://github.com/CGAL/cgal-public-dev/commit/a3431e3b3e half subsampled sphere under curvature flow example
https://github.com/CGAL/cgal-public-dev/commit/2bde8362e2 testing curvature flow, exploring robust degenerate handling, overloads and minor fixes
https://github.com/CGAL/cgal-public-dev/commit/5eede86eaa paste degenerate removal code in compatible remeshing class
https://github.com/CGAL/cgal-public-dev/commit/842301a60e more robust edge collapsing and use of repair.h for deg faces
https://github.com/CGAL/cgal-public-dev/commit/3cad7d0332 recent pull in repair.h - minor changes
https://github.com/CGAL/cgal-public-dev/commit/fafd857d92 remove degenerate edges and faces
https://github.com/CGAL/cgal-public-dev/commit/242c76bb6a curvature example in cmakelists
https://github.com/CGAL/cgal-public-dev/commit/86cbe29d70 plugin: dock widget design & support both compatible and curvature
https://github.com/CGAL/cgal-public-dev/commit/cbba9b9ba0 find new location using cot weigths
https://github.com/CGAL/cgal-public-dev/commit/1c2fddbdf9 normalize kn
https://github.com/CGAL/cgal-public-dev/commit/ad31962530 setup plugin for curvature flow
https://github.com/CGAL/cgal-public-dev/commit/6a26812b94 verbose info in smoothing.h and an overload
https://github.com/CGAL/cgal-public-dev/commit/54b7cee0bb add projection to initial surface on curvature smoothing
https://github.com/CGAL/cgal-public-dev/commit/2bf0e11cc0 mean curvature smoothing - testing with and without cot angles in simple cases
https://github.com/CGAL/cgal-public-dev/commit/63d3096199 calculate mean curvature flow with Monge jet fitting
https://github.com/CGAL/cgal-public-dev/commit/26f76e7f25 cotangent weights calculation
https://github.com/CGAL/cgal-public-dev/commit/b0de15fa30 add tests for invalid meshes and adjust tolerance a bit to comply with some extreme cases
https://github.com/CGAL/cgal-public-dev/commit/a21e0f6475 typo
https://github.com/CGAL/cgal-public-dev/commit/d32a887fed Squashed commit of the following:
https://github.com/CGAL/cgal-public-dev/commit/b280b622e4 compatible remeshing works via demo plugin
https://github.com/CGAL/cgal-public-dev/commit/7505201060 fix vector_property_map and vertex_point map
https://github.com/CGAL/cgal-public-dev/commit/a4c8cb76f0 basic setup for demo plugin - boost::vector_property_map to be fixed
https://github.com/CGAL/cgal-public-dev/commit/5933f85fed mean curvature function overload fix
https://github.com/CGAL/cgal-public-dev/commit/bb8b919c69 setup for curvature flow class
https://github.com/CGAL/cgal-public-dev/commit/0dd83b3c55 switch default angle weighting off temporarily
https://github.com/CGAL/cgal-public-dev/commit/7156d9e961 small fix for collapsed degenerate cases after first iteration when using angle weights and some typos
https://github.com/CGAL/cgal-public-dev/commit/5355836708 anneal learning rate asymptotically
https://github.com/CGAL/cgal-public-dev/commit/9da06a91ac anneal learning rate in gradient descent with a power law to avoid local minima
https://github.com/CGAL/cgal-public-dev/commit/c9e15cc2e6 take No_constraint_map from remesh_impl
https://github.com/CGAL/cgal-public-dev/commit/8454b28546 use (optionally) weighted angles on calculating the barycenter
https://github.com/CGAL/cgal-public-dev/commit/963b6af182 add gradient descent precision as named parameter
https://github.com/CGAL/cgal-public-dev/commit/a4a1234516 iterations and combination of area and angle remeshing
https://github.com/CGAL/cgal-public-dev/commit/92d48d8085 area: remove projection on the fly, projection at the end only
https://github.com/CGAL/cgal-public-dev/commit/55ecb4e75a add constrained edges map
https://github.com/CGAL/cgal-public-dev/commit/ccc9802919 merge area and angle classes
https://github.com/CGAL/cgal-public-dev/commit/59f023b2d7 new design: smoothing.h - internal/Isotropic_remeshing/smoothing_impl.h
https://github.com/CGAL/cgal-public-dev/commit/8d9d8fed89 projection in angle smoothing
https://github.com/CGAL/cgal-public-dev/commit/cc81dfb085 get rid of boilerplate in test cpp
https://github.com/CGAL/cgal-public-dev/commit/276a12a2af area - bugs and optimizations in gradient descent
https://github.com/CGAL/cgal-public-dev/commit/3bab1557f8 area smoothing tests
https://github.com/CGAL/cgal-public-dev/commit/685b990bda fix degenerate cases on angle smoothing
https://github.com/CGAL/cgal-public-dev/commit/1c5dc5c7e2 fix tolerance on 180 degrees degenerate case
https://github.com/CGAL/cgal-public-dev/commit/c51d4ba4f2 better approach to degenerate cases
https://github.com/CGAL/cgal-public-dev/commit/90fd22e924 fix nan degenerate case
https://github.com/CGAL/cgal-public-dev/commit/3a915afc2d handle degenerate cases and tests
https://github.com/CGAL/cgal-public-dev/commit/015dcc127a area smoothing - projection with tree constructed with copied triangles
https://github.com/CGAL/cgal-public-dev/commit/b92aa404b8 simple degenerate polygon
https://github.com/CGAL/cgal-public-dev/commit/5cbd8a60bf project to surface with an Incremental remesher
https://github.com/CGAL/cgal-public-dev/commit/a404641ab8 area smoothing 3D stable without projection
https://github.com/CGAL/cgal-public-dev/commit/8470a82a2d add on gitignore
https://github.com/CGAL/cgal-public-dev/commit/81e88ae423 simple 3D polygon
https://github.com/CGAL/cgal-public-dev/commit/b7b251c495 fix scaling of bisector to correct length
https://github.com/CGAL/cgal-public-dev/commit/89ebd55dc9 project to tangent plane
https://github.com/CGAL/cgal-public-dev/commit/3106305f82 angle-based smoothing
https://github.com/CGAL/cgal-public-dev/commit/da0b5a62d7 angle bisection
https://github.com/CGAL/cgal-public-dev/commit/f14c28d933 typo
https://github.com/CGAL/cgal-public-dev/commit/510a041aed setup simple example for basic angle smoothing

## More to do
Here is a [blog post](https://kkatrio.github.io/projects/smoothing-gsoc17/) with highlights and some results.

Things to do:
* Improve GUI demo functionality so that it enables smoothing for selected areas.
* Add edge flips and edge splits in compatible smoothing algorithm.
* Implement an implicit approach for the mean curvature flow algorithm and compare its results with the explicit one.
* Improve quality evaluation by calculating triangle skewness.
